src/radical/pilot/pmgr/launching/base.py

```diff
@@ -248,16 +249,20 @@ def _kill_pilots(self, pids):
         the request to get enacted, nor for it to arrive, but just send it.
         '''
 
-        if not pids or not self._pilots:
-            # nothing to do
-            return
+        if not pids:
+            if not self._pilots:
+                return
+            else:
+                pids = list(self._pilots.keys())
 
         with self._lock:
+
             for pid in pids:
 
                 self._log.debug('cancel pilot %s', pid)
                 if pid not in self._pilots:
                     self._log.warn('cannot cancel unknown pilot %s', pid)
```

Comment:
Can this really happen? Also, should this be an exception instead of a warning?

Reply:
It can happen in principle, but it is a very tight race.  Assume the application issues the following two statements:

```py
pilot = pmgr.submit_pilots(...)
pilot.cancel()
```

The two lines will create messages in two different channels: the submission goes via the normal ZMQ queues, the cancellation creates a command message on the control pubsub.  Since ordering between the two channels is undefined, the launcher could in principle receive the cancellation before ever launching the pilot.

Also, in a more complex code, one application thread could launch a pilot while another thread decides to terminate all pilots for some reason.  Since pilot uids can be manually assigned (hard coded), the threads would not even need to communicate uids and it is thus easy for them to go out of sync.

But thanks for asking, it's useful to double-check those assymptions!